core/src/main/java/io/substrait/relation/VirtualTableScan.java

```diff
@@ -14,11 +13,26 @@ public abstract class VirtualTableScan extends AbstractReadRel {
 
   public abstract List<Expression.StructLiteral> getRows();
 
+  /**
+   *
+   * <li>non-empty rowset
+   * <li>non-null field-names
+   * <li>no null rows
+   * <li>row shape must match field-list
+   */
+  @Value.Check
+  protected void check() {
```

Comment:
Just curious when is this called?

Reply:
https://immutables.github.io/immutable.html#precondition-check-method
Its called from the `validate` method in the generated `ImmutableVirtualTableScan`